modules/streamloader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/streamloader/Bigquery.scala

```diff
@@ -40,18 +44,37 @@ object Bigquery {
   def insert[F[_]: Sync](
     failedInsertProducer: PubsubProducer[F, FailedInsert],
     metrics: Metrics[F],
-    loaderRows: List[LoaderRow]
+    toLoad: List[LoaderRow]
   )(
     mkInsert: List[LoaderRow] => F[InsertAllResponse]
-  ): F[Unit] =
-    mkInsert(loaderRows).flatMap {
-      case response if response.hasErrors =>
-        val (failed, successful) = extractFailedAndSuccessfulRows(loaderRows, response)
-        handleFailedRows(metrics, failedInsertProducer, failed) *>
-          handleSuccessfulRows(metrics, successful)
-      case _ =>
-        handleSuccessfulRows(metrics, loaderRows)
-    }
+  ): F[Unit] = {
+
+    def go(toLoad: List[LoaderRow], prevFailures: List[LoaderRow], prevSuccesses: List[LoaderRow]): F[Unit] =
+      mkInsert(toLoad).flatMap { response =>
+        val partitioned = partitionRowResults(toLoad, response)
+        val logF = partitioned.reasonsAndLocations.toSeq.traverse {
+          case (reason, location) =>
+            // We cannot log the `message` because it contains customer data.  But `reason` and `location` are safe.
+            Logger[F].warn(s"Bigquery inserts failed. Reason: [$reason]. Location: [$location]")
+        }
+        if (partitioned.stopped.nonEmpty && partitioned.failed.nonEmpty) {
+          // The request contained valid rows, but they were stopped because the request also contained invalid rows.
+          // Iterate again without the invalid rows.
+          logF *> go(partitioned.stopped, partitioned.failed ::: prevFailures, partitioned.successful ::: prevSuccesses)
+        } else {
+          // Either:
+          // - All rows in this iteration were successful
+          // - All rows in this iteration were invalid
+          // - Some rows were stopped but there were no failures.
+          // The 3rd case is unexpected and unlikely. Here we treat stopped rows as failures, to avoid any potential deadlock.
+          logF *>
+            handleFailedRows(metrics, failedInsertProducer, partitioned.failed ::: partitioned.stopped ::: prevFailures) *>
```

Comment:
Would parallel make it a bit faster `(handleFailedRows, handleSuccessfulRows).parTupled`?